src/systems/acoustic_comms/AcousticComms.cc

```diff
@@ -0,0 +1,186 @@
+  return sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2));
+}
+
+AcousticComms::AcousticComms()
+  : dataPtr(gz::utils::MakeUniqueImpl<Implementation>())
+{
+}
+
+void AcousticComms::Load(
+    const Entity &_entity,
+    std::shared_ptr<const sdf::Element> _sdf,
+    EntityComponentManager &_ecm,
+    EventManager &_eventMgr)
+{
+  if (_sdf->HasElement("max_range"))
+  {
+    this->dataPtr->maxRange = _sdf->Get<double>("max_range");
+  }
+  if (_sdf->HasElement("speed_of_sound"))
+  {
+    this->dataPtr->speedOfSound = _sdf->Get<double>("speed_of_sound");
+  }
+}
+
+void AcousticComms::Step(
+    const UpdateInfo &_info,
+    const comms::Registry &_currentRegistry,
+    comms::Registry &_newRegistry,
+    EntityComponentManager &_ecm)
+{
+  // Initialize entity if needed.
+  for (auto & [address, content] : _currentRegistry)
+  {
+    if (content.entity == kNullEntity)
+    {
+      auto entities = entitiesFromScopedName(content.modelName, _ecm);
+      if (entities.empty())
+        continue;
+
+      auto entityId = *(entities.begin());
+      if (entityId == kNullEntity)
+        continue;
+
+      _newRegistry[address].entity = entityId;
+    }
+  }
+
+  for (auto & [address, content] : _currentRegistry)
+  {
+    // Reference to the outbound queue for this address.
+    auto &outbound = content.outboundMsgs;
+
+    // Is the source address bound?
+    auto itSrc = _currentRegistry.find(address);
+    bool srcAddressBound = itSrc != _currentRegistry.end();
+
+    // Is the source address attached to a model?
+    bool srcAddressAttachedToModel =
+      srcAddressBound && itSrc->second.entity != kNullEntity;
+
+    comms::DataQueue newOutbound;
+
+    if (srcAddressAttachedToModel)
+    {
+      // All these messages need to be processed.
+      for (auto &msg : outbound)
+      {
+        // Is the destination address bound?
+        auto itDst = _currentRegistry.find(msg->dst_address());
+        bool dstAddressBound = itDst != _currentRegistry.end();
+
+        // Is the destination address attached to a model?
+        bool dstAddressAttachedToModel =
+          dstAddressBound && itDst->second.entity != kNullEntity;
+
+        if (!dstAddressAttachedToModel)
+          continue;
+
+        // Calculate distance between the bodies.
+        auto poseSrc = worldPose(itSrc->second.entity, _ecm).Pos();
+        auto poseDst = worldPose(itDst->second.entity, _ecm).Pos();
+
+        auto distanceToTransmitter =
+          this->dataPtr->DistanceBetweenBodies(poseSrc,
+              poseDst);
+
+        // Calculate distance covered by the message.
+        std::chrono::steady_clock::time_point currTime(_info.simTime);
+        auto timeOfTransmission = msg->mutable_header()->stamp();
+
+        double currTimestamp =
+          currTime.time_since_epoch().count() / 1000000000.0;
+        double packetTimestamp =
+          static_cast<double>(timeOfTransmission.sec()) +
+          static_cast<double>(timeOfTransmission.nsec()) / 1000000000.0;
```

Comment:
@adityapande-1995 hmm, consider using `std::chrono` types wherever possible. Floating point epsilons can be problematic. 

```suggestion
        auto packetTimestamp =
          std::chrono::seconds(timeOfTransmission.sec()) +
          std::chrono::nanoseconds(timeOfTransmission.nsec());
```

Reply:
Done https://github.com/gazebosim/gz-sim/pull/1608/commits/8cb0a46f91b638ed3387697c2864afa499428156